### ISSUE TITLE
Unreachable code panic - possibly due to macro use looking like a variable declaration?

### ISSUE THREAD

**Issue by studoot** [rust-lang-nursery/rustfmt#1092]
I'm getting an unreachable code panic when attempting to format a rust file (which compiles successfully), reducing it down to the following offending code:

``` rust
#[macro_use]
extern crate nom;
use nom::IResult;
use std::str::from_utf8;

pub fn test(input: &[u8], max_size: usize) -> IResult<&[u8], &[u8]> {
    chain!(input, a:take!(max_size), || [])
}
```

I'm guessing it's probably something to do with rustfmt seeing `a:take!(max_size)` as a variable declaration or something, where that is nom's syntax for capturing the value of the `take!` parser, as changing that line to the code below eliminates the panic. Having said that, other uses of this syntax haven't caused `rustfmt` any problems.

```
    chain!(input, take!(max_size), || [])
```

The error I get (with backtrace, although as it's a release build, the backtrace probably isn't much use) is below:

```
C:\Users\u404261\Desktop>rustfmt bin.rs
thread '<main>' panicked at 'internal error: entered unreachable code', C:\Users\u404261\.cargo\registry\src\github.com-1ecc6299db9ec823\rustfmt-0.5.0\src\types.rs:590
stack backtrace:
   0:        0x1400ee162 - <unknown>
   1:        0x1400ed74d - <unknown>
   2:        0x1400dae3d - <unknown>
   3:        0x13fdeccf3 - <unknown>
   4:        0x13fe5e56c - <unknown>
   5:        0x13feaa502 - <unknown>
   6:        0x13fe47a04 - <unknown>
   7:        0x13fe9f67a - <unknown>
   8:        0x13fe9e854 - <unknown>
   9:        0x13fec4d12 - <unknown>
  10:        0x13fec5cb7 - <unknown>
  11:        0x13fe5520b - <unknown>
  12:        0x13fe47212 - <unknown>
  13:        0x13fe45a67 - <unknown>
  14:        0x13fe4bc8c - <unknown>
  15:        0x13fe3b061 - <unknown>
  16:        0x13fded61a - <unknown>
  17:        0x13fdf9fd7 - <unknown>
  18:        0x13fdff1b7 - <unknown>
  19:        0x13fde6652 - <unknown>
  20:        0x13fdeb31e - <unknown>
  21:        0x1400ed14c - <unknown>
  22:        0x1400f2e21 - <unknown>
  23:        0x1400ece84 - <unknown>
  24:        0x1400fe95e - <unknown>
  25:         0x770359bc - BaseThreadInitThunk
```

I'm using Windows 7, Rust 1.10.0 stable (x64, MSVC ABI), rustfmt 0.5.0 (also compiled with Rust 1.10.0 stable).

**Comment by paholg** [comment]
I have this same issue. For me, it crops up when using type macros.

Minimal example:
```rust
macro_rules! u {
    () => ( u32 );
}

fn main() {
    let x: u!() = 13;
}
```

Rust 1.13.0 beta, rustfmt 0.6.2.

**Comment by topecongiro** [comment]
Closed via #1604.
